collect_app/src/main/java/org/odk/collect/android/widgets/TimeWidget.java

```diff
@@ -15,295 +15,106 @@
 
-        if (getFormEntryPrompt().getAnswerValue() == null) {
-            clearAnswer();
+        if (prompt.getAnswerValue() == null) {
+            setTimeToCurrent();
+            binding.widgetAnswerText.setText(R.string.no_time_selected);
         } else {
             Date date = (Date) getFormEntryPrompt().getAnswerValue().getValue();
-
             DateTime dateTime = new DateTime(date);
-            updateTime(dateTime, true);
-        }
-    }
-
-    public int getHour() {
-        return hourOfDay;
-    }
 
-    public int getMinute() {
-        return minuteOfHour;
-    }
-
-    public boolean isNullAnswer() {
-        return nullAnswer;
-    }
+            hourOfDay = dateTime.getHourOfDay();
+            minuteOfHour = dateTime.getMinuteOfHour();
+            binding.widgetAnswerText.setText(DateTimeWidgetUtils.getTimeData(hourOfDay, minuteOfHour).getDisplayText());
+        }
 
-    public void setTimeToCurrent() {
-        updateTime(DateTime.now(), false);
+        return binding.getRoot();
     }
 
-    public void updateTime(DateTime dateTime) {
-        updateTime(dateTime, true);
+    @Override
+    public void clearAnswer() {
+        setTimeToCurrent();
+        binding.widgetAnswerText.setText(R.string.no_time_selected);
+        widgetValueChanged();
     }
 
-    public void updateTime(DateTime dateTime, boolean shouldUpdateLabel) {
-        updateTime(dateTime.getHourOfDay(), dateTime.getMinuteOfHour(), shouldUpdateLabel);
+    @Override
+    public IAnswerData getAnswer() {
+        return !binding.widgetAnswerText.getText().equals(getContext().getString(R.string.no_time_selected))
+                ? DateTimeWidgetUtils.getTimeData(hourOfDay, minuteOfHour)
+                : null;
     }
 
-    public void updateTime(int hourOfDay, int minuteOfHour, boolean shouldUpdateLabel) {
-        this.hourOfDay = hourOfDay;
-        this.minuteOfHour = minuteOfHour;
-
-        timePickerDialog.updateTime(hourOfDay, minuteOfHour);
-
-        if (shouldUpdateLabel) {
-            setTimeLabel();
-        }
+    @Override
+    public void setOnLongClickListener(OnLongClickListener l) {
+        binding.widgetButton.setOnLongClickListener(l);
+        binding.widgetAnswerText.setOnLongClickListener(l);
     }
 
     @Override
-    public void onTimeSet(TimePicker timePicker, int hourOfDay, int minute) {
-        timePicker.clearFocus();
-
-        this.hourOfDay = timePicker.getCurrentHour();
-        this.minuteOfHour = timePicker.getCurrentMinute();
-
-        setTimeLabel();
-        widgetValueChanged();
+    public void cancelLongPress() {
+        super.cancelLongPress();
+        binding.widgetButton.cancelLongPress();
+        binding.widgetAnswerText.cancelLongPress();
     }
 
     @Override
-    public void onButtonClick(int buttonId) {
-        if (nullAnswer) {
-            setTimeToCurrent();
-        } else {
-            updateTime(hourOfDay, minuteOfHour, true);
+    public void setBinaryData(Object answer) {
```

Comment:
As I said in one of my previous comments it would be good to rename this method because now it is used in many widgets not just binary ones. `setData` maybe?

Reply:
Oh okay right Sorry I forgot.